internal/http.go

```diff
@@ -0,0 +1,239 @@
+type HTTPError struct {
+	code    int
+	message string
+}
+
+func InitializeRouter(ctx context.Context) {
+	var reportLog = ctrl.Log.WithName("report")
+	http.HandleFunc("/download", downloadReport)
+
+	// viper.SetDefault("HTTP.PORT", ":5000")
+	httpServerPort := viper.Get("HTTP.PORT").(string)
+
+	server := http.Server{
+		Addr:        httpServerPort,
+		BaseContext: func(_ net.Listener) context.Context { return ctx },
+	}
+	reportLog.Info("starting http server on port" + httpServerPort)
+
+	go startServer(&server, reportLog)
+
+	stopCh, closeCh := signals.CreateChannel()
+	defer closeCh()
+	log.Println("system call notified:", <-stopCh)
+
+	shutdownServer(ctx, &server, reportLog)
+}
+
+func startServer(server *http.Server, reportLog logr.Logger) {
+	if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+		log.Fatal(err)
+	} else {
+		reportLog.Info("http server stopped gracefully")
+	}
+}
+
+func shutdownServer(ctx context.Context, server *http.Server, reportLog logr.Logger) {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	if err := server.Shutdown(ctx); err != nil {
+		reportLog.Error(err, "http server shutdown error:")
+	} else {
+		reportLog.Info("http server shutdowned")
+	}
+}
+
+func downloadReport(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	var reportLog = ctrl.Log.WithName("report")
+	HTTPError := validaeRequest(r, reportLog)
+	if HTTPError != nil {
+		http.Error(w, HTTPError.message, HTTPError.code)
+		return
+	}
+
+	conn, err := connectToDB(ctx)
+	if err != nil {
+		reportLog.Error(err, "database connection error:")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	err = generateReport(ctx, conn, r, reportLog)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	defer conn.Close(ctx)
+	w.WriteHeader(http.StatusOK)
+
+	_, err = w.Write([]byte("generated report"))
+	if err != nil {
+		reportLog.Error(err, "failed writing http response:")
+	}
+}
+
+func validaeRequest(r *http.Request, reportLog logr.Logger) *HTTPError {
+	if r.URL.Path != "/download" {
+		return &HTTPError{http.StatusNotFound, "404 not found"}
+	}
+
+	if r.Method != "GET" {
+		return &HTTPError{http.StatusMethodNotAllowed, "method is not supported"}
+	}
+
+	query := r.URL.Query()
+	start, present := query["start"]
+	if !present || len(start) == 0 {
+		reportLog.Info("start time not present")
+		return &HTTPError{http.StatusBadRequest, "start time not present"}
+	}
+
+	end, present := query["end"]
+	if !present || len(end) == 0 {
+		fmt.Println("end time not present")
+		return &HTTPError{http.StatusBadRequest, "end time not present"}
+	}
+	return nil
+}
+
+func connectToDB(ctx context.Context) (*pgx.Conn, error) {
+	host := viper.Get("DB.HOST").(string)
+	port := viper.GetInt("DB.PORT")
+	password := viper.Get("DB.PASSWORD").(string)
+	user := viper.Get("DB.USERNAME").(string)
+	dbname := viper.Get("DB.NAME").(string)
```

Comment:
All of these checks centered around deriving configuration values need to live in the main package.